staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

```diff
@@ -4609,6 +4623,19 @@ func (c *Cloud) getFullInstance(nodeName types.NodeName) (*awsInstance, *ec2.Ins
 	return awsInstance, instance, err
 }
 
+func (c *Cloud) nodeNameToProviderID(nodeName types.NodeName) (InstanceID, error) {
+	// TODO: get node from cache
+	node, err := c.kubeClient.CoreV1().Nodes().Get(string(nodeName), metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	awsID, err := KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
```

Comment:
To avoid #76976 again, I checked that `node.Spec.ProviderID` can be set only once and it's read-only after that:

https://github.com/kubernetes/kubernetes/blob/a3e914528a1c16af9797ff9e0e070a84e1842ba0/pkg/apis/core/validation/validation.go#L4208-L4215

A rogue node can't change it and do bad things except for initial registration of the node. @liggitt, @smarterclayton, is that safe enough?

Reply:
what would an invalid or malicious provider id set on registration allow a node to do? is it ever verified against metadata?

Reply:
@jsafrane @liggitt I checked node registration process, `ProviderID` is set during kubelet's initializing v1.Node, and kubelet gets this information from cloud provider directly: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L323

Given this, plus we only allow `ProviderID` to be set only once, I'd say its safe.

Reply:
> Though node controller does not double check, given it's reported by kubelet from cloud provider, plus we only allow `ProviderID` to be set only once, I'd say its safe enough?

In general, we need to correlate the information the node reports against something obtained from cloud provider metadata. If we're not going to do that, and accept what the node self-reports, we need to know the full impact of an incorrect or malicious report by a node.

Reply:
> we need to know the full impact of an incorrect or malicious report by a node

So I went through the code base about how node name and provider id is reported and used, and here is a summary:

1. Both nodeName and providerID are initially reported by kubelet.
2. Kubelet defaults nodeName to hostName, but will use the "nodeName" from cloud provider to override if possible
3. Kubelet reads providerID from cloud provider only if user does not specify it
4. Node controller will set providerID it got from cloud provider only when kubelet does not set it.
5. Node controller maintains a set of node informations (zone, region, instance type, node addresses) by node ID primarily, if it is not able to obtain information from cloud provider directly, it falls back to node name
6. Interface-wise, `cloudprovider.Instance` provides interfaces to query information both by nodeName and providerID and is used in various places

Given aws-specific context, defaulting getNodeByName to getNodeByProviderID shall be a good to have optimization for correct providerID, and with malicious providerID reporting (have to be done by user explicitly setting it in kubelet config), we have 2 cases:

- malicious id is mal-formated: `MapToAWSInstanceID()` will return error, and we fall back to query by name, which is safe, We waste 1 Get() call from api server for now but with #78409 the inefficiency would be minimum

-  malicious id is a valid provider ID, but is just not for the instance nodeName: given that node controller is already tracking information primarily using providerID, it's OK to me to keep the same assumption. But if we also want to prevent this scenario, we can start a different thread to discuss k8s's behavior regarding malicious providerID reporting.

@liggitt @jsafrane please suggest

Reply:
Rephrasing the above just to make sure I understood it, if a malicious node A sets ProviderID of another AWS machine (B) during node A registration, node controller will fill wrong IP addresses to Node A object and A/D controller will attach/detach volumes to machine B instead of real node A.

From storage perspective, Kubelet on node A won't see devices for the attached volumes and it won't start any pods that use the volumes. Dunno what else breaks - does something use `node.spec.addresses`?

The same can be achieved without this PR if node A registers a node with `node.name = B`. Do we have any checks that would prevent that?

Reply:
> The same can be achieved without this PR if node A registers a node with `node.name = B`. Do we have any checks that would prevent that?

Yes, NodeRestriction admission would not permit that.

Reply:
> malicious id is a valid provider ID, but is just not for the instance nodeName

this is the case I care the most about.

I had a hard time following all the paths you mentioned. Is this PR making the situation worse than it currently is? Are there places we would currently catch the mismatch that this PR makes us tolerate?

Reply:
@jsafrane 

> The same can be achieved without this PR if node A registers a node with node.name = B. Do we have any checks that would prevent that?

Theoretically node name can be faked. However in aws context, Node name coming from cloud provider will always override whatever value user sets / defaults to

>  if a malicious node A sets ProviderID of another AWS machine (B) during node A registration, node controller will fill wrong IP addresses to Node A object
if a wrong but valid providerID is reported, we will fill out wrong IP in most cases

> does something use node.spec.addresses
mainly 2 places - volume and pod with host networking

Reply:
@liggitt 

>  Is this PR making the situation worse than it currently is? Are there places we would currently catch the mismatch that this PR makes us tolerate?

This PR does not making things worse, and it improves performance in aws context
This PR does not add additional verification / break current verification of node name/ip mismatch in aws context.

Reply:
> This PR does not making things worse

IMO it does. Previously, controller-manager used `node.name` for all its decissions, which cannot be faked by a rogue node, as @liggitt wrote above.

With this PR, controller-manager uses `node.spec.providerID`, which is IMO not validated by API server and can be faked by a rogue node.

> This PR does not add additional verification / break current verification of node name/ip mismatch in aws context

That's correct, on the other hand it introduces possibility of `providerID` mismatch.

Reply:
> Previously, controller-manager used `node.name`

Not exactly as “cloudprovider.Instances” provides interface for get information from cloud provider by id, and this semantic opens up choice to track things by ID, which is what node controller is doing currently.